### PR TITLE
first cut of client

### PR DESCRIPTION
**Pull request by kevholditch**

### CODE REVIEW

File: pkg/vaultclient/client.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	Insecure bool
+}
+
+type Auth struct {
+	token  string
+	expiry time.Time
+}
+
+var (
+	expirationWindow = time.Duration(10) * time.Second
```
Comment (pawelkowalak) [inline pkg/vaultclient/client.go]: Can be replaced with `time.Second * 10`

File: pkg/vaultclient/client.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	return &iamAuth{
+		client: c,
+		role:   cfg.IamRole,
+	}, nil
+}
+
+func (v *iamAuth) getAuth() (*Auth, error) {
+
```
Comment (pawelkowalak) [inline pkg/vaultclient/client.go]: Remove blank line

File: pkg/vaultclient/client.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+			IamRole:  role,
+		}
+	}
+
+	return nil
+}
+
+func NewVaultAuth(cfg *Config) (VaultAuth, error) {
+	config := api.DefaultConfig()
+	err := config.ConfigureTLS(&api.TLSConfig{Insecure: cfg.Insecure})
```
Comment (pawelkowalak) [inline pkg/vaultclient/client.go]: `if err := config...; err != nil {`

File: pkg/vaultclient/client.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	if err != nil {
+		return nil, err
+	}
+	c, err := api.NewClient(config)
+
```
Comment (pawelkowalak) [inline pkg/vaultclient/client.go]: Remove blank line or move above line 76.

File: pkg/vaultclient/client.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+		return &Config{
+			AuthType: Iam,
+			IamRole:  role,
+		}
+	}
+
+	return nil
+}
+
+func NewVaultAuth(cfg *Config) (VaultAuth, error) {
```
Comment (pawelkowalak) [inline pkg/vaultclient/client.go]: We should return a struct, not interface. We can't be sure if caller is more interested in a concrete type or in behavior.

File: pkg/vaultclient/client.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+		client: c,
+		role:   cfg.IamRole,
+	}, nil
+}
+
+func (v *iamAuth) getAuth() (*Auth, error) {
+
+	s := session.Must(session.NewSession())
+	data, err := awsauth.GenerateLoginData(s.Config.Credentials, "")
+
```
Comment (pawelkowalak) [inline pkg/vaultclient/client.go]: Remove blank line.

File: pkg/vaultclient/client.go (hunk 5 above)
Comment (kevholditch) [reply pkg/vaultclient/client.go]: need to return interface as there are two implementations token auth and iam auth.  The caller supplies to config and then gets an implementation that suits them

Comment (pawelkowalak) [reply pkg/vaultclient/client.go]: 👍 didn't notice

**Pull request merged**